File: backend/app/core/urls.py
"""URL normalization, domain handling, and business/website identity matching."""

from __future__ import annotations

import re
from typing import Any, Dict, List, Optional, Set, Tuple
from urllib.parse import urljoin, urlparse, urlunparse

import tldextract

# Offline extractor: never phones home for the public-suffix list mid-crawl.
_extract = tldextract.TLDExtract(suffix_list_urls=())

TRACKING_PARAMS = {
    "utm_source", "utm_medium", "utm_campaign", "utm_term", "utm_content", "utm_id",
    "gclid", "fbclid", "msclkid", "mc_cid", "mc_eid", "ref", "referrer", "_ga", "yclid",
}

# Platforms that are a *profile*, not the business's own website. Treating a
# Facebook page as "their website" would poison the whole audit.
SOCIAL_HOSTS = {
    "facebook.com", "m.facebook.com", "fb.com", "fb.me", "instagram.com", "twitter.com",
    "x.com", "linkedin.com", "youtube.com", "youtu.be", "tiktok.com", "pinterest.com",
    "snapchat.com", "threads.net", "wa.me", "api.whatsapp.com", "t.me", "telegram.me",
}

DIRECTORY_HOSTS = {
    "yelp.com", "yellowpages.com", "yell.com", "tripadvisor.com", "trustpilot.com",
    "google.com", "goo.gl", "maps.app.goo.gl", "business.site", "bing.com",
    "foursquare.com", "angi.com", "angieslist.com", "houzz.com", "thumbtack.com",
    "checkatrade.com", "bark.com", "truelocal.com.au", "hotfrog.com", "manta.com",
    "bbb.org", "opentable.com", "doordash.com", "ubereats.com", "grubhub.com",
    "booksy.com", "fresha.com", "treatwell.com", "zocdoc.com", "healthgrades.com",
}

LINK_IN_BIO_HOSTS = {
    "linktr.ee", "linkin.bio", "beacons.ai", "carrd.co", "bio.link", "milkshake.app",
}

NON_WEBSITE_HOSTS = SOCIAL_HOSTS | DIRECTORY_HOSTS | LINK_IN_BIO_HOSTS

NON_HTML_EXT = {
    ".pdf", ".jpg", ".jpeg", ".png", ".gif", ".svg", ".webp", ".ico", ".zip", ".rar",
    ".mp4", ".mp3", ".avi", ".mov", ".doc", ".docx", ".xls", ".xlsx", ".ppt", ".pptx",
    ".css", ".js", ".json", ".xml", ".woff", ".woff2", ".ttf", ".eot", ".dmg", ".exe",
}

_LEGAL_SUFFIXES = {
    "llc", "inc", "ltd", "limited", "plc", "co", "corp", "corporation", "company",
    "gmbh", "pty", "pvt", "private", "llp", "lp", "sa", "srl", "bv", "nv", "ag", "ab",
    "oy", "as", "kft", "spa", "sl", "the", "and", "of",
}

_GENERIC_WORDS = {
    "services", "service", "solutions", "group", "center", "centre", "shop", "store",
    "studio", "clinic", "salon", "cafe", "restaurant", "bar", "hotel", "school",
    "academy", "agency", "consulting", "consultants", "partners", "associates",
}


# --------------------------------------------------------------------------
# Normalization
# --------------------------------------------------------------------------


def normalize_url(raw: str) -> Optional[str]:
    """Turn a messy CSV cell into a canonical absolute URL, or None."""
    if not raw:
        return None
    url = str(raw).strip().strip('"').strip("'")
    if not url or url.lower() in {"n/a", "na", "none", "null", "-", "no website", "nan"}:
        return None

    url = re.sub(r"\s+", "", url)
    if url.startswith("//"):
        url = "https:" + url
    if not re.match(r"^[a-zA-Z][a-zA-Z0-9+.\-]*://", url):
        if "." not in url.split("/")[0]:
            return None
        url = "https://" + url

    try:
        p = urlparse(url)
    except ValueError:
        return None
    if p.scheme not in ("http", "https") or not p.netloc:
        return None

    host = p.netloc.lower()
    if host.endswith(":80"):
        host = host[:-3]
    elif host.endswith(":443"):
        host = host[:-4]
    if "." not in host.split(":")[0]:
        return None

    path = p.path or "/"
    query = _strip_tracking(p.query)
    return urlunparse((p.scheme, host, path, "", query, ""))


def _strip_tracking(query: str) -> str:
    if not query:
        return ""
    kept = []
    for part in query.split("&"):
        if not part:
            continue
        key = part.split("=", 1)[0].lower()
        if key not in TRACKING_PARAMS:
            kept.append(part)
    return "&".join(kept)


def registrable_domain(url: str) -> str:
    if not url:
        return ""
    try:
        host = urlparse(url).netloc or url
    except ValueError:
        host = url
    host = host.split("@")[-1].split(":")[0].lower()
    ext = _extract(host)
    if ext.domain and ext.suffix:
        return f"{ext.domain}.{ext.suffix}"
    return host


def host_of(url: str) -> str:
    """Hostname without the port - used for grouping and comparisons."""
    try:
        return (urlparse(url).netloc or "").split(":")[0].lower()
    except ValueError:
        return ""


def netloc_of(url: str) -> str:
    """Hostname *with* the port, when one is present."""
    try:
        return (urlparse(url).netloc or "").lower()
    except ValueError:
        return ""


def origin_of(url: str) -> str:
    """
    scheme://host[:port] - the base for /robots.txt and /sitemap.xml.

    The port must be kept: dropping it sends those lookups to port 80 and makes
    every site on a non-standard port look as though it publishes neither file.
    """
    try:
        p = urlparse(url)
    except ValueError:
        return ""
    if not p.netloc:
        return ""
    return f"{(p.scheme or 'https').lower()}://{p.netloc.lower()}"


def same_site(a: str, b: str) -> bool:
    da, db = registrable_domain(a), registrable_domain(b)
    return bool(da) and da == db


def is_non_website_host(url: str) -> Tuple[bool, str]:
    """True when the URL is a social/directory profile rather than a real site."""
    dom = registrable_domain(url)
    host = host_of(url)
    if not dom:
        return False, ""
    if dom in SOCIAL_HOSTS or host in SOCIAL_HOSTS:
        return True, "social_profile"
    if dom in DIRECTORY_HOSTS or host in DIRECTORY_HOSTS:
        return True, "directory_listing"
    if dom in LINK_IN_BIO_HOSTS or host in LINK_IN_BIO_HOSTS:
        return True, "link_in_bio"
    return False, ""


def is_crawlable(url: str) -> bool:
    try:
        p = urlparse(url)
    except ValueError:
        return False
    if p.scheme not in ("http", "https"):
        return False
    path = (p.path or "").lower()
    return not any(path.endswith(ext) for ext in NON_HTML_EXT)


def absolutize(base: str, href: str) -> Optional[str]:
    if not href:
        return None
    href = href.strip()
    low = href.lower()
    if low.startswith(("javascript:", "mailto:", "tel:", "sms:", "data:", "#", "callto:", "whatsapp:")):
        return None
    try:
        joined = urljoin(base, href)
    except ValueError:
        return None
    joined = joined.split("#", 1)[0]
    return normalize_url(joined)


def url_key(url: str) -> str:
    """Dedup key: ignores scheme, leading www, trailing slash, and index files."""
    if not url:
        return ""
    try:
        p = urlparse(url)
    except ValueError:
        return url
    host = (p.netloc or "").lower().replace("www.", "", 1).split(":")[0]
    path = re.sub(r"/(index|default|home)\.(html?|php|aspx?)$", "/", p.path or "/", flags=re.I)
    path = path.rstrip("/") or "/"
    q = f"?{p.query}" if p.query else ""
    return f"{host}{path}{q}".lower()


# --------------------------------------------------------------------------
# Identity matching (spec 7) - never attach another company's website
# --------------------------------------------------------------------------


def name_tokens(name: str) -> List[str]:
    n = (name or "").lower()
    n = re.sub(r"[&+]", " and ", n)
    n = re.sub(r"[^a-z0-9]+", " ", n)
    toks = [t for t in n.split() if t and t not in _LEGAL_SUFFIXES and len(t) > 1]
    return toks


def distinctive_tokens(name: str) -> List[str]:
    toks = name_tokens(name)
    distinct = [t for t in toks if t not in _GENERIC_WORDS]
    return distinct or toks


def _domain_core(url: str) -> str:
    dom = registrable_domain(url)
    ext = _extract(dom)
    return re.sub(r"[^a-z0-9]", "", (ext.domain or dom).lower())


def score_identity(
    *,
    business_name: str,
    url: str,
    page_title: str = "",
    page_text: str = "",
    phone_digits: Optional[List[str]] = None,
    city: str = "",
    postal_code: str = "",
    address: str = "",
    category: str = "",
) -> Dict[str, Any]:
    """
    Confidence 0..1 that `url` really belongs to `business_name`.

    Weighted signals, each independently evidenced so the UI can explain the
    verdict rather than showing a bare number.
    """
    signals: List[Dict[str, Any]] = []
    score = 0.0

    toks = name_tokens(business_name)
    distinct = distinctive_tokens(business_name)
    core = _domain_core(url)
    text_l = (page_text or "").lower()
    title_l = (page_title or "").lower()

    # 1. Domain contains the business name tokens (strongest single signal).
    if core and distinct:
        joined = "".join(distinct)
        initials = "".join(t[0] for t in distinct if t)
        if joined and joined in core:
            score += 0.45
            signals.append({"signal": "domain_matches_full_name", "weight": 0.45, "detail": core})
        else:
            hits = [t for t in distinct if len(t) >= 4 and t in core]
            if hits:
                frac = len(hits) / len(distinct)
                gain = 0.34 * frac
                score += gain
                signals.append(
                    {"signal": "domain_matches_name_tokens", "weight": round(gain, 3),
                     "detail": ",".join(hits)}
                )
            elif len(initials) >= 3 and initials in core:
                score += 0.16
                signals.append({"signal": "domain_matches_initials", "weight": 0.16, "detail": initials})

    # 2. Business name appears in the page title.
    if title_l and distinct:
        hits = [t for t in distinct if t in title_l]
        if hits and len(hits) / len(distinct) >= 0.6:
            score += 0.22
            signals.append({"signal": "name_in_title", "weight": 0.22, "detail": page_title[:120]})
        elif hits:
            score += 0.10
            signals.append({"signal": "partial_name_in_title", "weight": 0.10, "detail": ",".join(hits)})

    # 3. Business name appears in page body.
    if text_l and distinct:
        hits = [t for t in distinct if t in text_l]
        if len(hits) == len(distinct):
            score += 0.14
            signals.append({"signal": "name_in_page_text", "weight": 0.14, "detail": "all tokens"})
        elif hits:
            score += 0.07
            signals.append({"signal": "partial_name_in_page_text", "weight": 0.07, "detail": ",".join(hits)})

    # 4. The CSV phone number appears on the site - very strong corroboration.
    if phone_digits:
        page_digits = re.sub(r"\D", "", text_l)
        for pd in phone_digits:
            tail = re.sub(r"\D", "", pd)[-9:]
            if len(tail) >= 7 and tail in page_digits:
                score += 0.28
                signals.append({"signal": "phone_found_on_site", "weight": 0.28, "detail": pd})
                break

    # 5. Location corroboration.
    if postal_code and len(postal_code.strip()) >= 4:
        pc = postal_code.strip().lower().replace(" ", "")
        if pc and pc in text_l.replace(" ", ""):
            score += 0.12
            signals.append({"signal": "postal_code_on_site", "weight": 0.12, "detail": postal_code})
    if city and len(city) > 2 and city.lower() in text_l:
        score += 0.08
        signals.append({"signal": "city_on_site", "weight": 0.08, "detail": city})
    if address:
        street = re.sub(r"[^a-z0-9 ]", " ", address.lower())
        parts = [p for p in street.split() if len(p) > 3][:4]
        if parts and sum(1 for p in parts if p in text_l) >= max(2, len(parts) - 1):
            score += 0.10
            signals.append({"signal": "address_on_site", "weight": 0.10, "detail": address[:120]})

    # 6. Weak category corroboration.
    if category:
        cat_toks = [t for t in re.split(r"[^a-z]+", category.lower()) if len(t) > 3]
        if cat_toks and any(t in text_l for t in cat_toks):
            score += 0.05
            signals.append({"signal": "category_language_present", "weight": 0.05, "detail": category})

    confidence = round(min(1.0, score), 3)
    if confidence >= 0.8:
        verdict = "strong_match"
    elif confidence >= 0.55:
        verdict = "probable_match"
    elif confidence >= 0.3:
        verdict = "weak_match"
    else:
        verdict = "no_match"

    return {"confidence": confidence, "verdict": verdict, "signals": signals}


def candidate_domains(business_name: str, tlds: Optional[List[str]] = None) -> List[str]:
    """
    Conservative domain guesses for businesses whose CSV row has no website.

    A guess is only ever a *candidate* - it must still pass identity matching
    against the fetched page before it is accepted (spec 7).
    """
    distinct = distinctive_tokens(business_name)
    if not distinct:
        return []
    tlds = tlds or [".com"]
    joined = "".join(distinct)
    hyphen = "-".join(distinct)

    bases: List[str] = []
    if 3 <= len(joined) <= 30:
        bases.append(joined)
    if len(distinct) > 1 and 3 <= len(hyphen) <= 34:
        bases.append(hyphen)
    if len(distinct) > 2:
        short = "".join(distinct[:2])
        if 3 <= len(short) <= 30:
            bases.append(short)

    out: List[str] = []
    seen: Set[str] = set()
    for base in bases:
        for tld in tlds:
            d = f"https://{base}{tld}"
            if d not in seen:
                seen.add(d)
                out.append(d)
    return out[:6]


COUNTRY_TLDS = {
    "US": [".com", ".net"],
    "CA": [".ca", ".com"],
    "GB": [".co.uk", ".com", ".uk"],
    "AU": [".com.au", ".com", ".au"],
    "NZ": [".co.nz", ".com"],
    "IE": [".ie", ".com"],
    "PK": [".com.pk", ".pk", ".com"],
    "IN": [".in", ".co.in", ".com"],
    "DE": [".de", ".com"],
    "FR": [".fr", ".com"],
    "NL": [".nl", ".com"],
    "ES": [".es", ".com"],
    "IT": [".it", ".com"],
    "ZA": [".co.za", ".com"],
    "AE": [".ae", ".com"],
}


def tlds_for_region(region: Optional[str]) -> List[str]:
    return COUNTRY_TLDS.get((region or "").upper(), [".com", ".net"])

File: cloudflare/src/core/urls.ts
// URL normalization, domain handling, and business/website identity matching.
// Ported from backend/app/core/urls.py — logic and thresholds unchanged.
import { getDomain } from "tldts";

const TRACKING_PARAMS = new Set([
  "utm_source", "utm_medium", "utm_campaign", "utm_term", "utm_content", "utm_id",
  "gclid", "fbclid", "msclkid", "mc_cid", "mc_eid", "ref", "referrer", "_ga", "yclid",
]);

const SOCIAL_HOSTS = new Set([
  "facebook.com", "m.facebook.com", "fb.com", "fb.me", "instagram.com", "twitter.com",
  "x.com", "linkedin.com", "youtube.com", "youtu.be", "tiktok.com", "pinterest.com",
  "snapchat.com", "threads.net", "wa.me", "api.whatsapp.com", "t.me", "telegram.me",
]);

const DIRECTORY_HOSTS = new Set([
  "yelp.com", "yellowpages.com", "yell.com", "tripadvisor.com", "trustpilot.com",
  "google.com", "goo.gl", "maps.app.goo.gl", "business.site", "bing.com",
  "foursquare.com", "angi.com", "angieslist.com", "houzz.com", "thumbtack.com",
  "checkatrade.com", "bark.com", "truelocal.com.au", "hotfrog.com", "manta.com",
  "bbb.org", "opentable.com", "doordash.com", "ubereats.com", "grubhub.com",
  "booksy.com", "fresha.com", "treatwell.com", "zocdoc.com", "healthgrades.com",
]);

const LINK_IN_BIO_HOSTS = new Set(["linktr.ee", "linkin.bio", "beacons.ai", "carrd.co", "bio.link", "milkshake.app"]);

const NON_HTML_EXT = new Set([
  ".pdf", ".jpg", ".jpeg", ".png", ".gif", ".svg", ".webp", ".ico", ".zip", ".rar",
  ".mp4", ".mp3", ".avi", ".mov", ".doc", ".docx", ".xls", ".xlsx", ".ppt", ".pptx",
  ".css", ".js", ".json", ".xml", ".woff", ".woff2", ".ttf", ".eot", ".dmg", ".exe",
]);

const LEGAL_SUFFIXES = new Set([
  "llc", "inc", "ltd", "limited", "plc", "co", "corp", "corporation", "company",
  "gmbh", "pty", "pvt", "private", "llp", "lp", "sa", "srl", "bv", "nv", "ag", "ab",
  "oy", "as", "kft", "spa", "sl", "the", "and", "of",
]);

const GENERIC_WORDS = new Set([
  "services", "service", "solutions", "group", "center", "centre", "shop", "store",
  "studio", "clinic", "salon", "cafe", "restaurant", "bar", "hotel", "school",
  "academy", "agency", "consulting", "consultants", "partners", "associates",
]);

export function normalizeUrl(raw: string | null | undefined): string | null {
  if (!raw) return null;
  let url = String(raw).trim().replace(/^"|"$/g, "").replace(/^'|'$/g, "");
  if (!url || ["n/a", "na", "none", "null", "-", "no website", "nan"].includes(url.toLowerCase())) return null;

  url = url.replace(/\s+/g, "");
  if (url.startsWith("//")) url = "https:" + url;
  if (!/^[a-zA-Z][a-zA-Z0-9+.\-]*:\/\//.test(url)) {
    if (!url.split("/")[0].includes(".")) return null;
    url = "https://" + url;
  }

  let p: URL;
  try {
    p = new URL(url);
  } catch {
    return null;
  }
  if (p.protocol !== "http:" && p.protocol !== "https:" || !p.host) return null;

  let host = p.host.toLowerCase();
  if (host.endsWith(":80")) host = host.slice(0, -3);
  else if (host.endsWith(":443")) host = host.slice(0, -4);
  if (!host.split(":")[0].includes(".")) return null;

  const path = p.pathname || "/";
  const query = stripTracking(p.search.replace(/^\?/, ""));
  return `${p.protocol}//${host}${path}${query ? "?" + query : ""}`;
}

function stripTracking(query: string): string {
  if (!query) return "";
  const kept: string[] = [];
  for (const part of query.split("&")) {
    if (!part) continue;
    const key = part.split("=", 1)[0].toLowerCase();
    if (!TRACKING_PARAMS.has(key)) kept.push(part);
  }
  return kept.join("&");
}

export function registrableDomain(url: string | null | undefined): string {
  if (!url) return "";
  let host: string;
  try {
    host = new URL(url).host || url;
  } catch {
    host = url;
  }
  host = host.split("@").pop()!.split(":")[0].toLowerCase();
  return getDomain(host, { allowPrivateDomains: true }) || host;
}

export function hostOf(url: string): string {
  try {
    return (new URL(url).host || "").split(":")[0].toLowerCase();
  } catch {
    return "";
  }
}

export function originOf(url: string): string {
  try {
    const p = new URL(url);
    if (!p.host) return "";
    return `${(p.protocol || "https:").toLowerCase()}//${p.host.toLowerCase()}`;
  } catch {
    return "";
  }
}

export function sameSite(a: string, b: string): boolean {
  const da = registrableDomain(a);
  const db = registrableDomain(b);
  return Boolean(da) && da === db;
}

export function isNonWebsiteHost(url: string): { isProfile: boolean; kind: string } {
  const dom = registrableDomain(url);
  const host = hostOf(url);
  if (!dom) return { isProfile: false, kind: "" };
  if (SOCIAL_HOSTS.has(dom) || SOCIAL_HOSTS.has(host)) return { isProfile: true, kind: "social_profile" };
  if (DIRECTORY_HOSTS.has(dom) || DIRECTORY_HOSTS.has(host)) return { isProfile: true, kind: "directory_listing" };
  if (LINK_IN_BIO_HOSTS.has(dom) || LINK_IN_BIO_HOSTS.has(host)) return { isProfile: true, kind: "link_in_bio" };
  return { isProfile: false, kind: "" };
}

export function isCrawlable(url: string): boolean {
  let p: URL;
  try {
    p = new URL(url);
  } catch {
    return false;
  }
  if (p.protocol !== "http:" && p.protocol !== "https:") return false;
  const path = (p.pathname || "").toLowerCase();
  for (const ext of NON_HTML_EXT) if (path.endsWith(ext)) return false;
  return true;
}

export function absolutize(base: string, href: string | null | undefined): string | null {
  if (!href) return null;
  href = href.trim();
  const low = href.toLowerCase();
  if (["javascript:", "mailto:", "tel:", "sms:", "data:", "#", "callto:", "whatsapp:"].some((p) => low.startsWith(p))) {
    return null;
  }
  let joined: string;
  try {
    joined = new URL(href, base).toString();
  } catch {
    return null;
  }
  joined = joined.split("#")[0];
  return normalizeUrl(joined);
}

export function urlKey(url: string): string {
  if (!url) return "";
  let p: URL;
  try {
    p = new URL(url);
  } catch {
    return url;
  }
  const host = (p.host || "").toLowerCase().replace(/^www\./, "").split(":")[0];
  let path = (p.pathname || "/").replace(/\/(index|default|home)\.(html?|php|aspx?)$/i, "/");
  path = path.replace(/\/+$/, "") || "/";
  const q = p.search ? p.search : "";
  return `${host}${path}${q}`.toLowerCase();
}

// -- identity matching ------------------------------------------------------

export function nameTokens(name: string): string[] {
  let n = (name || "").toLowerCase();
  n = n.replace(/[&+]/g, " and ");
  n = n.replace(/[^a-z0-9]+/g, " ");
  return n.split(/\s+/).filter((t) => t && !LEGAL_SUFFIXES.has(t) && t.length > 1);
}

export function distinctiveTokens(name: string): string[] {
  const toks = nameTokens(name);
  const distinct = toks.filter((t) => !GENERIC_WORDS.has(t));
  return distinct.length ? distinct : toks;
}

function domainCore(url: string): string {
  const dom = registrableDomain(url);
  const ext = getDomain(dom, { allowPrivateDomains: true });
  const label = (ext || dom).split(".")[0] || dom;
  return label.toLowerCase().replace(/[^a-z0-9]/g, "");
}

export interface IdentitySignal {
  signal: string;
  weight: number;
  detail: string;
}

export interface IdentityScore {
  confidence: number;
  verdict: "strong_match" | "probable_match" | "weak_match" | "no_match";
  signals: IdentitySignal[];
}

export function scoreIdentity(opts: {
  businessName: string;
  url: string;
  pageTitle?: string;
  pageText?: string;
  phoneDigits?: string[];
  city?: string;
  postalCode?: string;
  address?: string;
  category?: string;
}): IdentityScore {
  const signals: IdentitySignal[] = [];
  let score = 0;

  const distinct = distinctiveTokens(opts.businessName);
  const core = domainCore(opts.url);
  const textL = (opts.pageText || "").toLowerCase();
  const titleL = (opts.pageTitle || "").toLowerCase();

  if (core && distinct.length) {
    const joined = distinct.join("");
    const initials = distinct.map((t) => t[0]).join("");
    if (joined && core.includes(joined)) {
      score += 0.45;
      signals.push({ signal: "domain_matches_full_name", weight: 0.45, detail: core });
    } else {
      const hits = distinct.filter((t) => t.length >= 4 && core.includes(t));
      if (hits.length) {
        const gain = 0.34 * (hits.length / distinct.length);
        score += gain;
        signals.push({ signal: "domain_matches_name_tokens", weight: round(gain, 3), detail: hits.join(",") });
      } else if (initials.length >= 3 && core.includes(initials)) {
        score += 0.16;
        signals.push({ signal: "domain_matches_initials", weight: 0.16, detail: initials });
      }
    }
  }

  if (titleL && distinct.length) {
    const hits = distinct.filter((t) => titleL.includes(t));
    if (hits.length && hits.length / distinct.length >= 0.6) {
      score += 0.22;
      signals.push({ signal: "name_in_title", weight: 0.22, detail: (opts.pageTitle || "").slice(0, 120) });
    } else if (hits.length) {
      score += 0.1;
      signals.push({ signal: "partial_name_in_title", weight: 0.1, detail: hits.join(",") });
    }
  }

  if (textL && distinct.length) {
    const hits = distinct.filter((t) => textL.includes(t));
    if (hits.length === distinct.length) {
      score += 0.14;
      signals.push({ signal: "name_in_page_text", weight: 0.14, detail: "all tokens" });
    } else if (hits.length) {
      score += 0.07;
      signals.push({ signal: "partial_name_in_page_text", weight: 0.07, detail: hits.join(",") });
    }
  }

  if (opts.phoneDigits?.length) {
    const pageDigits = textL.replace(/\D/g, "");
    for (const pd of opts.phoneDigits) {
      const tail = pd.replace(/\D/g, "").slice(-9);
      if (tail.length >= 7 && pageDigits.includes(tail)) {
        score += 0.28;
        signals.push({ signal: "phone_found_on_site", weight: 0.28, detail: pd });
        break;
      }
    }
  }

  if (opts.postalCode && opts.postalCode.trim().length >= 4) {
    const pc = opts.postalCode.trim().toLowerCase().replace(/\s+/g, "");
    if (pc && textL.replace(/\s+/g, "").includes(pc)) {
      score += 0.12;
      signals.push({ signal: "postal_code_on_site", weight: 0.12, detail: opts.postalCode });
    }
  }
  if (opts.city && opts.city.length > 2 && textL.includes(opts.city.toLowerCase())) {
    score += 0.08;
    signals.push({ signal: "city_on_site", weight: 0.08, detail: opts.city });
  }
  if (opts.address) {
    const street = opts.address.toLowerCase().replace(/[^a-z0-9 ]/g, " ");
    const parts = street.split(/\s+/).filter((p) => p.length > 3).slice(0, 4);
    if (parts.length && parts.filter((p) => textL.includes(p)).length >= Math.max(2, parts.length - 1)) {
      score += 0.1;
      signals.push({ signal: "address_on_site", weight: 0.1, detail: opts.address.slice(0, 120) });
    }
  }
  if (opts.category) {
    const catToks = opts.category.toLowerCase().split(/[^a-z]+/).filter((t) => t.length > 3);
    if (catToks.length && catToks.some((t) => textL.includes(t))) {
      score += 0.05;
      signals.push({ signal: "category_language_present", weight: 0.05, detail: opts.category });
    }
  }

  const confidence = round(Math.min(1, score), 3);
  const verdict = confidence >= 0.8 ? "strong_match" : confidence >= 0.55 ? "probable_match" : confidence >= 0.3 ? "weak_match" : "no_match";
  return { confidence, verdict, signals };
}

export function candidateDomains(businessName: string, tlds: string[] = [".com"]): string[] {
  const distinct = distinctiveTokens(businessName);
  if (!distinct.length) return [];
  const joined = distinct.join("");
  const hyphen = distinct.join("-");

  const bases: string[] = [];
  if (joined.length >= 3 && joined.length <= 30) bases.push(joined);
  if (distinct.length > 1 && hyphen.length >= 3 && hyphen.length <= 34) bases.push(hyphen);
  if (distinct.length > 2) {
    const short = distinct.slice(0, 2).join("");
    if (short.length >= 3 && short.length <= 30) bases.push(short);
  }

  const out: string[] = [];
  const seen = new Set<string>();
  for (const base of bases) {
    for (const tld of tlds) {
      const d = `https://${base}${tld}`;
      if (!seen.has(d)) {
        seen.add(d);
        out.push(d);
      }
    }
  }
  return out.slice(0, 6);
}

const COUNTRY_TLDS: Record<string, string[]> = {
  US: [".com", ".net"], CA: [".ca", ".com"], GB: [".co.uk", ".com", ".uk"],
  AU: [".com.au", ".com", ".au"], NZ: [".co.nz", ".com"], IE: [".ie", ".com"],
  PK: [".com.pk", ".pk", ".com"], IN: [".in", ".co.in", ".com"], DE: [".de", ".com"],
  FR: [".fr", ".com"], NL: [".nl", ".com"], ES: [".es", ".com"], IT: [".it", ".com"],
  ZA: [".co.za", ".com"], AE: [".ae", ".com"],
};

export function tldsForRegion(region?: string | null): string[] {
  return COUNTRY_TLDS[(region || "").toUpperCase()] || [".com", ".net"];
}

function round(n: number, digits: number): number {
  const f = 10 ** digits;
  return Math.round(n * f) / f;
}
